src/rateEngine/billingDeterminants/SurchargeAsPercent.ts

```diff
@@ -14,6 +13,10 @@ class SurchargeAsPercent extends BillingDeterminants {
   constructor({rateElement}: SurchargeAsPercentArgs) {
     super();
 
+    if (!rateElement) {
+      throw new Error("Cannot calculate surcharge without a rate element.")
```

Comment:
I think maybe this is trying to tell us that `rateElement` should be optional here:

https://github.com/bellawatt/electric-rate-engine/blob/f741cdfbba045e058e2d2359935d3c2bfdc724e9/src/rateEngine/types/index.ts#L203

What do you think?

Reply:
ok, i'd like to take a closer look at this one

Reply:
Ok, so what's going on is that their are actually 2 different data shapes for `SurchargeAsPercent` rate elements: 

1. One that the user defines, which is shorthand for
2. A second one (that includes a `rateElement` property) which exists after the input data is processed through the `RateComponentsFactory`.

Making `rateElement` optional in `SurchargeAsPercentArgs` was a hack to get around that (that I must have made way back when). The newly added strict type checking is now surfacing this to us.

Throwing an error when rateElement is undefined would also be papering over what's essentially a single type trying to describe two distinct types.

Working on a fix still.

Reply:
This should do it!

https://github.com/dcordz/electric-rate-engine/pull/2

Reply:
Thanks for looking into that, I didn't really feel comfortable doing too much here although I guess this PR did get away a little bit from its original intent/scope.

Reply:
Classic programmer's credo:

![image](https://github.com/user-attachments/assets/48eae17d-7dfd-4158-8d63-568cf89a4ea7)

In all seriousness, this is exactly what I'd hope to find by changing to strict type checking. It's uncovered a few small flaws in how we built our types and code. It's definitely not easy to reason about code that's been sitting around for a few years, but certainly in scope for this kind of library optimization/improvement PR!